### ISSUE TITLE
strpos() Error in PHP 7.3

### ISSUE THREAD

**Issue by kchevalier-mg** [timber/timber#1899]
Hi Guys. First off, thanks for working on this great project. We've incorporated it into our base WordPress theme for many of our new projects, you can see that base theme here: https://github.com/mindgruve/mg-press

I've encountered an issue when testing on PHP 7.3. Running a `resize` filter on an image `src` property throws a Twig_Error_Runtime exception:

```
An exception has been thrown during the rendering of a template ("strpos(): Non-string needles will be interpreted as strings in the future. Use an explicit chr() call to preserve the current behavior").
```

The Twig markup looks like this:

```
<img src="{{ post.thumbnail.src|resize(1200, 300) }}">
```

And the stack trace points to line 457 (strpos) in file `timber-library/lib/ImageHelper.php`:

```
    protected static function is_in_theme_dir( $path ) {
        $root = realpath(get_stylesheet_directory_uri());
        if ( 0 === strpos($path, $root) ) {
            return true;
        }
    }
```

`get_stylesheet_directory_uri()` should return a URI, and I'm pretty sure that `realpath()` is intended to work with filepaths only, so $root is always `null`, and now PHP 7.3 is complaining. This looks like a bug. If I switch that to `get_stylesheet_directory()` instead it works as expected.

### Expected behavior
Resized image.

### Actual behavior
Twig_Error_Runtime exception thrown.

### Steps to reproduce behavior
Add to Twig template:

```
<img src="{{ post.thumbnail.src|resize(1200, 300) }}">
```

Upgrade to PHP 7.3.

### What version of WordPress, PHP and Timber are you using?
WordPress 5.0.3, PHP 7.3, Timber 1.8.4

### How did you install Timber? (for example, from GitHub, Composer/Packagist, WP.org?)
Composer

**Comment by albertonieto** [comment]
Same issue here, please advice.

**Comment by jarednova** [comment]
Thanks @albertonieto and @kchevalier — at this point I've only used PHP 7.1 for testing. I'll switch to PHP 7.3 to see if I can replicate issue

**Comment by palmiak** [comment]
@albertonieto @kchevalier I added a PR to fix this problem

**Comment by fluff** [comment]
That PR does not look right at all. `chr()` returns a single character based on an ascii code, but that's not what you want to `strpos` search into. 

I think one of the reasons strpos gives a warning here, at least for me, is that `get_stylesheet_directory_uri()` returns an absolute URL (starting with https:// and all) and `realpath()` returns false because it obviously can't parse URLs, and so strpos gets a boolean $needle parameter and that causes this warning.

**Comment by palmiak** [comment]
Hi @fluff you're right - I misread `chr` absolutly. Thanks for spotting this.

I think the best way would be just to add anothe `if` to do something when we get `false`:
```
protected static function is_in_theme_dir( $path ) {
    $root = realpath(get_stylesheet_directory_uri());

    if ( false === $root ) {
        return false;
    }

    if ( 0 === strpos($path, (string)$root) ) {
        return true;
    }
}
```

this `string` is a bit overprotection but to be sure $root will treated as one.

**Comment by kchevalier** [comment]
What is the point of `realpath(get_stylesheet_directory_uri())`? When will it ever return something other than `false`?

`realpath()` returns an absolute path, while `get_stylesheet_directory_uri()` returns a URI. Unless I'm mistaken, those two are not compatible.

I think the answer is to drop `_uri` and use `get_stylesheet_directory()` instead.

**Comment by palmiak** [comment]
good point @kchevalier I also was thinking about this. There is https://github.com/timber/timber/blob/3beeae4ccb7310e17de04555573b0948873a73b4/tests/test-timber-image.php#L746 - so this `realpath(uri)` is also used in test. 

So either @jarednova had some very specific case in mind that we can't think of or we can try do as you propose.

**Comment by jarednova** [comment]
It all seems a part of this commit:

https://github.com/timber/timber/commit/0164f2ddc760a419522726e57c182bf7d5e9fd51

... seems to have something to do with situations in which `.../` appears in a path to a theme. If that's the blocking part of this issue, I think PHP 7.3 compatibility is probably more important than whatever edge case this attempts to solve

**Comment by albertonieto** [comment]
I have updated to the latest version (1.9.0), but the issue still persists with php 7.3, ideas?

**Comment by palmiak** [comment]
@albertonieto It wasn't yet fixed.

For a quick fix you can use:
```
protected static function is_in_theme_dir( $path ) {
    $root = realpath(get_stylesheet_directory_uri());

    if ( false === $root ) {
        return false;
    }

    if ( 0 === strpos($path, (string)$root) ) {
        return true;
    }
}
```

It's not perfect, but it will work. I hope we'll fix this in this week.

**Comment by albertonieto** [comment]
Thanks, indeed I was using your code as a temp fix. I hope it gets released soon. Thanks for the help.

**Comment by palmiak** [comment]
@albertonieto this is already merged to master so you can try to see if it's correct.

**Comment by cyphix333** [comment]
Just came here to report the same thing after upgrading to PHP 7.3, glad to see it's being worked on :)

**Comment by cyphix333** [comment]
@kchevalier-mg How did you manage to get a `Twig_Error_Runtime exception thrown.`? The only reason I seen the error was because the images were broken (the deprecated notice was hidden as it was stuck inside the image attribute).

**Comment by okebinda** [comment]
@cyphix333 (This is kchevalier-mg on my personal account, not work) We use Whoops error handling. Here's how it is set up in the project: https://github.com/mindgruve/mg-press/blob/master/lib/errors.php

**Comment by cyphix333** [comment]
@okebinda Thanks a lot for that :+1:

**Comment by jarednova** [comment]
Closed by #1915 and released as Timber 1.9.1
